### PR TITLE
Reworked sequences

### PR DESCRIPTION
**Pull request by leostryuk**
### What does this PR do?

Sequences now are accessed, incremented, created in separate transactions.
It was required to support contract for xodus-dnq.

### Motivation

What inspired you to submit this pull request?

### Related issues

A list of issues either fixed, containing architectural discussions, otherwise relevant
for this Pull Request.

### Additional Notes

Anything else we should know when reviewing?

### Checklist

[] I have added unit tests to the code I am submitting
[] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture

### CODE REVIEW

File: entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/iterate/property/OSequenceImpl.kt [hunk 1]
```diff
@@ -24,36 +25,46 @@ import jetbrains.exodus.entitystore.Sequence
...
 
-
-    private fun getOrCreateSequence(start:Long): OSequence {
-        val session = ODatabaseSession.getActiveSession()
-        val sequenceLibrary: OSequenceLibrary = session.metadata.sequenceLibrary
-        var result = sequenceLibrary.getSequence(sequenceName)
-        if (result == null) {
-            val params = CreateParams().setStart(start).setIncrement(1)
-            result = sequenceLibrary.createSequence(sequenceName, SEQUENCE_TYPE.ORDERED, params)
+    private fun <T> accessSequenceInSession(action: (OSequence) -> T): T {
```
Comment (vladimir-zatsepin) [inline entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/iterate/property/OSequenceImpl.kt:51]:
```suggestion
    private fun <T> withSequenceInSession(action: (OSequence) -> T): T {
```

File: entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/iterate/property/OSequenceImpl.kt [hunk 2]
```diff
@@ -24,36 +25,46 @@ import jetbrains.exodus.entitystore.Sequence
...
+                val params = CreateParams().setStart(initialValue).setIncrement(1)
+                oSequence = sequenceLibrary.createSequence(sequenceName, SEQUENCE_TYPE.ORDERED, params)
+                session.commit()
+                session.begin()
+            }
+            val actionResult = action(oSequence)
+            session.commit()
+            actionResult
         }
+        currentSession?.activateOnCurrentThread()
```
Comment (vladimir-zatsepin) [inline entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/iterate/property/OSequenceImpl.kt:67]: Lets move it into `finally` block

File: entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/OPersistentEntityStore.kt [hunk 3]
```diff
@@ -65,7 +65,7 @@ class OPersistentEntityStore(
         val txn = session.begin().transaction
         session.activateOnCurrentThread()
 
-        currentTx = OStoreTransactionImpl(session, txn, this, schemaBuddy)
+        currentTx = OStoreTransactionImpl(session, txn, this, schemaBuddy, databaseProvider::acquireSession)
```
Comment (vladimir-zatsepin) [inline entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/OPersistentEntityStore.kt:68]: Do we need `session` as param here?

**Review by vladimir-zatsepin — APPROVED** [review]

File: entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/OPersistentEntityStore.kt (hunk 3 above)
Comment (leostryuk) [reply entity-store/src/main/kotlin/jetbrains/exodus/entitystore/orientdb/OPersistentEntityStore.kt:68]: we need. sometimes it's session created before and managed by PersistentStore

**Pull request merged**